pstoreds/addr_book.go

```diff
@@ -2,398 +2,608 @@ package pstoreds
+func (ab *dsAddrBook) loadRecord(id peer.ID, cache bool, update bool) (pr *addrsRecord, err error) {
+	if e, ok := ab.cache.Get(id); ok {
+		pr = e.(*addrsRecord)
+		if pr.Refresh() && update {
+			ab.asyncFlush(pr)
 		}
-		log.Errorf("failed to delete addresses for peer %s: %s\n", p.Pretty(), err)
+		return pr, nil
 	}
 
+	txn, err := ab.ds.NewTransaction(true)
 	if err != nil {
-		log.Errorf("failed to avoid write conflict for peer %s after %d retries: %v\n", p.Pretty(), mgr.writeRetries, err)
-		return err
+		return nil, err
 	}
+	defer txn.Discard()
 
-	return nil
-}
+	key := addrBookBase.ChildString(b32.RawStdEncoding.EncodeToString([]byte(id)))
+	data, err := txn.Get(key)
 
-func (mgr *dsAddrBook) setAddrs(p peer.ID, addrs []ma.Multiaddr, ttl time.Duration, mode ttlWriteMode) error {
-	// Keys and cleaned up addresses.
-	keys, addrs, err := keysAndAddrs(p, addrs)
-	if err != nil {
-		return err
+	if err != nil && err != ds.ErrNotFound {
+		return nil, err
 	}
 
-	mgr.cache.Remove(p)
-	// Attempt transactional KV insertion.
-	var existed []bool
-	for i := 0; i < mgr.writeRetries; i++ {
-		if existed, err = mgr.dbInsert(keys, addrs, ttl, mode); err == nil {
-			break
+	if err == nil {
+		pr = arPool.Get().(*addrsRecord)
+		if err = pr.Unmarshal(data); err != nil {
+			return nil, err
+		}
+		if pr.Refresh() && update {
+			ab.asyncFlush(pr)
 		}
-		log.Errorf("failed to write addresses for peer %s: %s\n", p.Pretty(), err)
+	} else {
+		pr = arPool.Get().(*addrsRecord)
+		pr.Id = &pb.ProtoPeerID{ID: id}
 	}
 
-	if err != nil {
-		log.Errorf("failed to avoid write conflict for peer %s after %d retries: %v\n", p.Pretty(), mgr.writeRetries, err)
-		return err
+	if cache {
+		ab.cache.Add(id, pr)
 	}
+	return pr, nil
+}
+
+// background runs the housekeeping process that takes care of:
+//
+// * GCing expired addresses from the datastore at regular intervals.
+// * persisting asynchronous flushes to the datastore.
+func (ab *dsAddrBook) background() {
+	// placeholder tickers.
+	pruneTimer, lookaheadTimer := new(time.Ticker), new(time.Ticker)
+
+	// populate the tickers after the initial delay has passed.
+	go func() {
+		select {
+		case <-time.After(ab.opts.GCInitialDelay):
+			pruneTimer = time.NewTicker(ab.opts.GCPurgeInterval)
+			lookaheadTimer = time.NewTicker(ab.opts.GCLookaheadInterval)
+		}
+	}()
+
+	for {
+		select {
+		case fj := <-ab.flushJobCh:
+			if cached, ok := ab.cache.Peek(fj.Id.ID); ok {
+				// Only continue flushing if the record we have in memory is the same as for which the flush
+				// job was requested. If it's not in memory, it has been evicted and we don't know if we hold
+				// the latest state or not. Similarly, if it's cached but the pointer is different, it means
+				// it was evicted and has been reloaded, so we're also uncertain if we hold the latest state.
+				if pr := cached.(*addrsRecord); pr == fj {
+					pr.RLock()
+					pr.Flush(ab.ds)
```

Comment:
Can we somehow batch here to avoid a bunch of small writes (and to make this a bit more concurrent).

Reply:
I can remove the TX semantics and switch to autobatch, how does that sound?

Reply:
See https://github.com/libp2p/go-libp2p-peerstore/pull/47#discussion_r237942984. I can batch these within transactions manually, committing based on a ticker and on maximum outstanding operations (whichever comes first). However, keeping the transaction pending for a time, we do expose ourselves to transaction failures more often (if other writes come in between).

Reply:
And we _really_ should avoid that because strictly speaking, we are not a transaction (i.e. we don't need atomicity for writes inside a single txn).

Reply:
Done in the latest commit. Autobatch won't work for us because it's defined for datastores that have no native mechanism for batching.